File: Sets.rb

require 'BitSet'

class Sets

  def Sets.Empty(s)
    s.each do |x| 
      return false if (x) 
    end
    return true
  end
	
  def Sets.Different(s1, s2)
    return s1 != s2
#    max = s1.size
#    for i in 0..(max-1)
#      return false if (s1[i] == s2[i])
#    end
#    return true
  end
  
  def Sets.Includes(s1, s2)
    r = true
    for i in 0...s2.size
      if (s2[i] != s1[i]) then
	r = false
	break
      end
    end
    return r
  end
  
  def Sets.FullSet(max)
    s = BitSet.new
    for i in 0..max
      s.set(i)
    end
    return s
  end
	
  def Sets.Size(s)
    return s.trueCount

#    size = 0
#    max = s.size
#    for i in 0..(max-1)
#      size += 1 if (s[i]) 
#    end
#    return size
  end
	
  def Sets.First(s)
    r = -1
    for i in 0...s.size
      if (s[i]) then
	r = i
	break
      end
    end
    return r
  end
	
  def Sets.Differ(s, s1)
    max = s.size
    for i in 0..(max-1)
      s.clear(i) if (s1[i])
    end
  end
end

File: Sets.java
package Coco;

import java.util.BitSet;

class Sets {

	static boolean Empty(BitSet s) {			/** s=={}? */
	    boolean r = true;
		int max = s.size();
		for (int i=0; i<=max && r; i++) {
			if (s.get(i)) r = false;
		}
		return r;
	}
	
	static boolean Different(BitSet s1, BitSet s2) {	/** s1*s2=={}? */
	    boolean r = true;
		int max = s1.size();
		for (int i=0; i<=max && r; i++)
			if (s1.get(i) && s2.get(i)) r = false;
		return r;
	}
	
	static boolean Includes(BitSet s1, BitSet s2) {	/** s1 > s2 ? */
	    boolean r = true;
		int max = s2.size();
		for (int i=0; i<=max && r; i++)
			if (s2.get(i) && !s1.get(i)) r = false;
		return r;
	}
	
	static BitSet FullSet(int max) {			/** return {0..max} */
		BitSet s = new BitSet();
		for (int i=0; i<=max; i++) s.set(i);
		return s;
	}
	
	static int Size(BitSet s) {					/** return number of elements in s */
		int size = 0, max = s.size();
		for (int i=0; i<=max; i++)
			if (s.get(i)) size++;
		return size;
	}
	
	static int First(BitSet s) {				/** return first element in s */
	    int r = -1;
		int max = s.size();
		for (int i=0; i<=max && r < 0; i++)
			if (s.get(i)) r = i;
		return r;
	}
	
	static void Differ(BitSet s, BitSet s1) {	/** s = s - s1 */
		int max = s.size();
		for (int i=0; i<=max; i++) {
			if (s1.get(i)) s.clear(i);
		}
	}
	
}
